sdks/java/io/tika/src/main/java/org/apache/beam/sdk/io/tika/TikaSource.java

```diff
@@ -0,0 +1,466 @@
+import org.apache.beam.sdk.io.Source;
+import org.apache.beam.sdk.io.fs.MatchResult;
+import org.apache.beam.sdk.io.fs.MatchResult.Metadata;
+import org.apache.beam.sdk.options.PipelineOptions;
+import org.apache.tika.config.TikaConfig;
+import org.apache.tika.exception.TikaException;
+import org.apache.tika.io.TikaInputStream;
+import org.apache.tika.parser.AutoDetectParser;
+import org.apache.tika.parser.ParseContext;
+import org.apache.tika.parser.Parser;
+import org.joda.time.Instant;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.xml.sax.SAXException;
+import org.xml.sax.helpers.DefaultHandler;
+
+/**
+ * Implementation detail of {@link TikaIO.Read}.
+ *
+ * <p>A {@link Source} which can represent the content of the files parsed by Apache Tika.
+ */
+class TikaSource extends BoundedSource<String> {
+  private static final long serialVersionUID = -509574062910491122L;
+  private static final Logger LOG = LoggerFactory.getLogger(TikaSource.class);
+
+  @Nullable
+  private MatchResult.Metadata singleFileMetadata;
+  private final Mode mode;
+  private final TikaIO.Read spec;
+
+  /**
+   * Source mode.
+   */
+  public enum Mode {
+    FILEPATTERN, SINGLE_FILE
+  }
+
+  TikaSource(TikaIO.Read spec) {
+    this.mode = Mode.FILEPATTERN;
+    this.spec = spec;
+  }
+
+  TikaSource(Metadata fileMetadata, TikaIO.Read spec) {
+    mode = Mode.SINGLE_FILE;
+    this.singleFileMetadata = checkNotNull(fileMetadata, "fileMetadata");
+    this.spec = spec;
+  }
+
+  @Override
+  public BoundedReader<String> createReader(PipelineOptions options) throws IOException {
+    this.validate();
+    checkState(spec.getFilepattern().isAccessible(),
+        "Cannot create a Tika reader without access to the file"
+        + " or pattern specification: {}.", spec.getFilepattern());
+    if (spec.getTikaConfigPath() != null) {
+      checkState(spec.getTikaConfigPath().isAccessible(),
+        "Cannot create a Tika reader without access to its configuration",
+        spec.getTikaConfigPath());
+    }
+
+    String fileOrPattern = spec.getFilepattern().get();
+    if (mode == Mode.FILEPATTERN) {
+      List<Metadata> fileMetadata = expandFilePattern(fileOrPattern);
+      List<TikaReader> fileReaders = new ArrayList<>();
+      for (Metadata metadata : fileMetadata) {
+        fileReaders.add(new TikaReader(this, metadata.resourceId().toString()));
+      }
+      if (fileReaders.size() == 1) {
+        return fileReaders.get(0);
+      }
+      return new FilePatternTikaReader(this, fileReaders);
+    } else {
+      return new TikaReader(this, singleFileMetadata.resourceId().toString());
+    }
+
+  }
+
+  @Override
+  public List<? extends TikaSource> split(long desiredBundleSizeBytes, PipelineOptions options)
+    throws Exception {
+    if (mode == Mode.SINGLE_FILE) {
+      return ImmutableList.of(this);
+    } else {
+      List<Metadata> fileMetadata = expandFilePattern(spec.getFilepattern().get());
```

Comment:
Seems like there's no splitting going on. This definitely should have been a ParDo.

Reply:
Unless I'm mistaken it would then block the thread, Tika Parsers have no way of doing what TextIO can...